### ISSUE TITLE
New version of htsjdk

### ISSUE THREAD

**Issue by SergejN** [igvteam/igv#843]
Dear maintainers,

please, replace the htsjdk library 2.21.3 with the newest release 2.23.0 as it supports CSI files instead of BAI files, which is vital when working with large genomes such as that of the Mexican axolotl.

thank you!

**Comment by jrobinso** [comment]
CSi have been supported for well over a year.

**Comment by SergejN** [comment]
That's what I also saw in the change logs. However, today I have a bam file of the axolotl data and IGV fails to load the file although `samtools tview` works just file. I though it might be related to somewhat outdated version of the htsjdk. However, you posted while I was typing an update - I patched IGV and replaced htsjdk by the current version. However, IGV still refuses to load that file. So far I tested samtools and the UCSC genome browser - both are fine with that file.. I don't know what else I can try..

**Comment by jrobinso** [comment]
If you can create a small test file to reproduce the problem, or give me access to one of your files I will try it.    Are there any errors in the igv.log file?

**Comment by SergejN** [comment]
yes, sure. I extracted chr10p from the genome and the reads mapping to that chromosome arm from the bam file. BAM files were sorted using `samtools sort` (v1.10) and indexed with `samtools index -c -m 14`. 
On my laptop I get this error message
![image](https://user-images.githubusercontent.com/10404050/91876775-67841d00-ec7d-11ea-8131-e3407231dbc6.png)

Interestingly, when I tried to reproduce the behavior with the test data it worked fine if the data were loaded from a local file, but failed if the data were fetched through a URL. Is there is difference in what IGV expects to get if the data are fetched from a remote server?
Top track is the test data, while the bottom track is the full dataset. 
![image](https://user-images.githubusercontent.com/10404050/91877496-5687db80-ec7e-11ea-8c5d-6bb32a97e683.png)
Unfortunately, the server is only reachable from within our intranet, therefore, I cannot share the link.
I put the archive with the test data to Dropbox https://www.dropbox.com/sh/dsgj4fe0gpej4jm/AADWmPPzQnOGkSnL6SUx3Ceja?dl=0
In that archive, the index file is renamed to `bai`, although it's a `csi` file. This is because I tested it with the genome browser (it still needs the file to be called `*.bai`, but excepts `csi` files) to make sure the behavior didn't change with the test dataset, but forgot to rename it back.

this is what I see in the IGV output when I load the local file (first two lines) vs remote file
```
INFO [2020-09-01T18:20:00,903]  [IGV.java:1332] [pool-3-thread-1]  Loading 1 resources.
INFO [2020-09-01T18:20:00,906]  [TrackLoader.java:122] [pool-3-thread-1]  Loading resource, path /private/tmp/igv/testdata/possorted_genome_bam43.bam
INFO [2020-09-01T18:20:18,882]  [IGV.java:1332] [pool-3-thread-2]  Loading 1 resources.
INFO [2020-09-01T18:20:18,882]  [TrackLoader.java:122] [pool-3-thread-2]  Loading resource, path https://tanaka.static.vbc.ac.at/genomebrowser/data/Nastya/127143/possorted_genome_bam43.bam
ERROR [2020-09-01T18:20:19,007]  [AlignmentTileLoader.java:313] [pool-2-thread-2]  Error loading alignment data
htsjdk.samtools.util.RuntimeIOException: Invalid file header in BAM index https://tanaka.static.vbc.ac.at/genomebrowser/data/Nastya/127143/possorted_genome_bam43.bam.bai: �
	at htsjdk.samtools.AbstractBAMFileIndex.verifyIndexMagicNumber(AbstractBAMFileIndex.java:378) ~[htsjdk-2.23.0.jar:?]
	at htsjdk.samtools.AbstractBAMFileIndex.<init>(AbstractBAMFileIndex.java:70) ~[htsjdk-2.23.0.jar:?]
	at htsjdk.samtools.AbstractBAMFileIndex.<init>(AbstractBAMFileIndex.java:56) ~[htsjdk-2.23.0.jar:?]
	at htsjdk.samtools.DiskBasedBAMFileIndex.<init>(DiskBasedBAMFileIndex.java:42) ~[htsjdk-2.23.0.jar:?]
	at htsjdk.samtools.BAMFileReader.getIndex(BAMFileReader.java:416) ~[htsjdk-2.23.0.jar:?]
	at htsjdk.samtools.BAMFileReader.createIndexIterator(BAMFileReader.java:952) ~[htsjdk-2.23.0.jar:?]
	at htsjdk.samtools.BAMFileReader.query(BAMFileReader.java:612) ~[htsjdk-2.23.0.jar:?]
	at htsjdk.samtools.SamReader$PrimitiveSamReaderToSamReaderAdapter.query(SamReader.java:533) ~[htsjdk-2.23.0.jar:?]
	at htsjdk.samtools.SamReader$PrimitiveSamReaderToSamReaderAdapter.query(SamReader.java:538) ~[htsjdk-2.23.0.jar:?]
	at org.broad.igv.sam.reader.BAMReader.query(BAMReader.java:180) ~[igv.jar:?]
	at org.broad.igv.sam.AlignmentTileLoader.loadTile(AlignmentTileLoader.java:160) [igv.jar:?]
	at org.broad.igv.sam.AlignmentDataManager.loadInterval(AlignmentDataManager.java:405) [igv.jar:?]
	at org.broad.igv.sam.AlignmentDataManager.load(AlignmentDataManager.java:351) [igv.jar:?]
	at org.broad.igv.sam.CoverageTrack.load(CoverageTrack.java:175) [igv.jar:?]
	at org.broad.igv.ui.IGV.lambda$repaint$8(IGV.java:2388) [igv.jar:?]
	at java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
INFO [2020-09-01T18:20:19,008]  [MessageUtils.java:76] [pool-2-thread-2]  <html>Error encountered querying alignments: htsjdk.samtools.util.RuntimeIOException: Invalid file header in BAM index https://tanaka.static.vbc.ac.at/genomebrowser/data/Nastya/127143/possorted_genome_bam43.bam.bai: �
```

**Comment by jrobinso** [comment]
I'm not sure I understand the comment about the extension.   IGV (and the htsjdk) will not work if the csi file has a "bai" index,  the extension is used to determine the index type.   That would result in the error you see  "Invalid file header in BAM index".   If this file is in fact a "csi" index its not going to work with a "bai" extension
```
htsjdk.samtools.util.RuntimeIOException: Invalid file header in BAM index https://tanaka.static.vbc.ac.at/genomebrowser/data/Nastya/127143/possorted_genome_bam43.bam.bai
```

**Comment by SergejN** [comment]
I meant that the UCSC genome browser does not accept the extension `csi`. Therefore, in order to make the files work with that genome browser, I had to rename the file and forgot to rename it back. 
The problem also occurs if the csi file has the proper extension.
![image](https://user-images.githubusercontent.com/10404050/91897308-5d244c00-ec9a-11ea-8e7c-928a10702472.png)

**Comment by jrobinso** [comment]
OK I'll look at your test data later today.

**Comment by jrobinso** [comment]
The file is fine, as you have confirmed yourself, so this would indicate there is a problem with your server.   Perhaps it is not correctly supporting range requests.   Its also possible your server is returning an incorrect MIME type, we have seen this problem occasionally with Apache servers.    If your server is Apache see the note at the bottom of this page:  https://github.com/igvteam/igv.js/wiki/Data-Server-Requirements

**Comment by SergejN** [comment]
I don't think so. I made another test to check that.
I generated a BAM file that only contains coordinates <500,000,000 to be able to index it using `samtools index`:
`samtools view -b -h possorted_genome_bam43.bam chr10p:1-500000000 > test.bam; samtools index test.bam`
then I loaded the original BAM file from the server, the original BAM file from a different folder on the server to make sure it's not related to any weird permissions set by the software that generated the BAM file, and finally the modified small BAM file from the server. As you can see in the screenshot below, only the tracks that try to load the CSI file crash, but not the one with the BAI index suggesting that the server handles range requests correctly.

![image](https://user-images.githubusercontent.com/10404050/91982819-1599d080-ed2b-11ea-9e59-cba0cbfdeff6.png)

**Comment by SergejN** [comment]
I don't think so. I made another test to check that.
I generated a BAM file that only contains coordinates <500,000,000 to be able to index it using samtools index:
samtools view -b -h possorted_genome_bam43.bam chr10p:1-500000000 > test.bam; samtools index test.bam
then I loaded the original BAM file from the server, the original BAM file from a different folder on the server to make sure it's not related to any weird permissions set by the software that generated the BAM file, and finally the modified small BAM file from the server. As you can see in the screenshot below, only the tracks that try to load the CSI file crash, but not the one with the BAI index suggesting that the server handles range requests correctly.

![image](https://user-images.githubusercontent.com/10404050/91983914-59410a00-ed2c-11ea-8f87-abff286cf781.png)

P.S. I put in the wrong screenshot before, where the error message was wrong, but you could see that the reads are correctly displayed (and hence fetched from the server) in the track "test.bam". In the new screenshot the message is correct, but I missed the exact locus where the reads are.. Anyway, the last track is handled correctly. And that's the one with the BAI index

**Comment by jrobinso** [comment]
OK, in that case we will have to open an issue with the htsjdk team (https://github.com/samtools/htsjdk).   They will want a test case they can reproduce.   It might take me a little time to set that up,  if you open an issue there please CC me.

**Comment by SergejN** [comment]
I've done some more digging.. It seems that the problem has been reported before (see #827 ), but in a different context.

Further digging identified the following piece of code in htsjdk, which looks like it's trying to load the BAI file no matter what and throws an exception otherwise.

```
    protected void verifyIndexMagicNumber(final String sourceName) {
        // Verify the magic number.
        seek(0);
        final byte[] buffer = new byte[4];
        readBytes(buffer);
        if (!Arrays.equals(buffer, BAMFileConstants.BAI_INDEX_MAGIC)) {
            throw new RuntimeIOException("Invalid file header in BAM index " + sourceName +
                    ": " + new String(buffer));
        }
    }
```
https://github.com/samtools/htsjdk/blob/edb8c602bbc20ff268e56282163f65911eabe379/src/main/java/htsjdk/samtools/AbstractBAMFileIndex.java#L372

The question is - since the file name displayed in the error message is correct, why does htsjdk think it's a bai index? It obviously fails in that method.. but I cannot figure out why it happens

**Comment by jrobinso** [comment]
I'm closing this here, as it has to be solved in the htsjdk and there is already an issue open for that. https://github.com/samtools/htsjdk/issues/1507

**Comment by jrobinso** [comment]
Hi all, this has been fixed in the htsjdk and IGV (version 2.13.1).  Apologies for the long delay.
